File: slimjim49j/day_1/zig/day_1.zig
const std = @import("std");

const digits = [_][*:0]const u8{ "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    var file = try std.fs.cwd().openFile("input", .{});
    defer file.close();

    const fileSize = try file.getEndPos();
    const input = try allocator.alloc(u8, fileSize);
    _ = try file.readAll(input);

    var sum: usize = 0;
    var seenFirst = false;
    var first: usize = 0;
    var last: usize = 0;
    for (0..fileSize) |i| {
        const ch = input[i];
        if (ch >= 48 and ch <= 57) {
            const n = ch - 48;
            if (!seenFirst) {
                first = n;
                seenFirst = true;
            }
            last = n;
        } else if (ch >= 97 and ch <= 122) {
            for (0..digits.len) |j| {
                const digit = digits[j];
                var isMatch = true;
                var k: usize = 0;
                while (digit[k] != 0) {
                    const srcCh = input[i + k];
                    const cmpCh = digit[k];
                    if (srcCh != cmpCh) {
                        isMatch = false;
                        break;
                    }
                    k += 1;
                }
                if (isMatch) {
                    if (!seenFirst) {
                        first = j;
                        seenFirst = true;
                    }
                    last = j;
                    break;
                }
            }
        } else if (ch == 10) {
            sum += (first * 10) + last;
            seenFirst = false;
            first = 0;
            last = 0;
        }
    }

    std.debug.print("{d}\n", .{sum});
}

File: slimjim49j/day_1/rust/day_1.rs
use std::fs;

const DIGITS: &[&[u8]; 10] = &[
    "zero".as_bytes(),
    "one".as_bytes(),
    "two".as_bytes(),
    "three".as_bytes(),
    "four".as_bytes(),
    "five".as_bytes(),
    "six".as_bytes(),
    "seven".as_bytes(),
    "eight".as_bytes(),
    "nine".as_bytes(),
];

fn main() {
    let file = fs::read_to_string("./input").unwrap();
    let input = file.as_bytes();

    let mut sum = 0;
    let mut seen_first = false;
    let mut first: usize = 0;
    let mut last: usize = 0;
    for i in 0..input.len() {
        let ch = input[i] as usize;

        if ch >= 48 && ch <= 57 {
            let n: usize = ch - 48;
            if !seen_first {
                first = n;
                seen_first = true;
            }
            last = n;
        } else if ch >= 97 && ch <= 122 {
            for j in 0..DIGITS.len() {
                let digit = DIGITS[j];
                let mut is_match = true;
                for k in 0..digit.len() {
                    let src_ch = input[i + k];
                    let cmp_ch = digit[k];
                    if src_ch != cmp_ch {
                        is_match = false;
                        break;
                    }
                }
                if is_match {
                    if !seen_first {
                        first = j;
                        seen_first = true;
                    }
                    last = j;
                    break;
                }
            }
        } else if ch == 10 {
            sum += (first * 10) + last;
            seen_first = false;
            first = 0;
            last = 0;
        }
    }
    println!("{}", sum);
}
